machine_test.go

```diff
@@ -797,21 +819,90 @@ func TestCaptureFifoToFile(t *testing.T) {
 		}
 	}()
 
-	fifoLogPath := fifoPath + ".log"
-	fifo, err := os.OpenFile(fifoLogPath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
+	logPath := fifoPath + ".log"
+	logFile, err := os.OpenFile(logPath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
 	if err != nil {
 		t.Fatalf("Failed to create fifo file: %v", err)
 	}
 
-	if err := captureFifoToFile(fctesting.NewLogEntry(t), fifoPath, fifo); err != nil {
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	testWriter := &fctesting.TestWriter{
+		WriteFn: func(b []byte) (int, error) {
+			defer wg.Done()
+
+			return logFile.Write(b)
+		},
+	}
+
+	if err := captureFifoToFile(fctesting.NewLogEntry(t), fifoPath, testWriter); err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+
+	defer os.Remove(logPath)
+
+	wg.Wait()
+	_, err = os.Stat(logPath)
+	assert.NoError(t, err, "failed to stat file")
+	b, err := ioutil.ReadFile(logPath)
+	assert.NoError(t, err, "failed to read logPath")
+	assert.Equal(t, expectedBytes, b)
+}
+
+func TestCaptureFifoToFile_nonblock(t *testing.T) {
+	fifoPath := filepath.Join(testDataPath, "TestCaptureFifoToFile_nonblock")
+
+	if err := syscall.Mkfifo(fifoPath, 0700); err != nil {
+		t.Fatalf("Unexpected error during syscall.Mkfifo call: %v", err)
+	}
+	defer os.Remove(fifoPath)
+
+	logPath := fifoPath + ".log"
+	logFile, err := os.OpenFile(logPath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
+	if err != nil {
+		t.Fatalf("Failed to create log file: %v", err)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	testWriter := &fctesting.TestWriter{
+		WriteFn: func(b []byte) (int, error) {
+			defer wg.Done()
+
+			return logFile.Write(b)
+		},
+	}
+
+	if err := captureFifoToFile(fctesting.NewLogEntry(t), fifoPath, testWriter); err != nil {
 		t.Errorf("Unexpected error: %v", err)
 	}
 
-	defer os.Remove(fifoLogPath)
+	defer os.Remove(logPath)
 
-	if _, err := os.Stat(fifoLogPath); err != nil {
-		t.Errorf("Failed to stat file: %v", err)
+	f, err := os.OpenFile(fifoPath, os.O_RDWR, 0600)
+	if err != nil {
+		t.Fatalf("Failed to open file, %q: %v", fifoPath, err)
 	}
+	expectedBytes := []byte("Hello world!")
+	f.Write(expectedBytes)
+	defer f.Close()
+
+	go func() {
```

Comment:
nit: Can you change this to: `time.AfterFunc(250*time.Millisecond, func() { f.Close() })`?  https://golang.org/pkg/time/#AfterFunc

Reply:
Fixed

Reply:
You don't need the goroutine and you don't need to drain `t.C`:

> AfterFunc waits for the duration to elapse and then calls f in its own goroutine. It returns a Timer that can be used to cancel the call using its Stop method.

Reply:
Sorry, thought you just wanted to swap the lines. Fixed.